### PR TITLE
Validate some types with CI

### PR DESCRIPTION
**Pull request by michaelfig**
This moves types around (as per #1311), improves some, with minor code changes to conform, and uses `tsc` to validate Zoe and ERTP against those types.

### CODE REVIEW

File: packages/ERTP/src/issuer.js
Comment (dckc) [inline packages/ERTP/src/issuer.js:20]: I see types such as `Issuer` got moved to `types.js`; how is the reader of `issuer.js` expected to discover this? Just by convention? I'd prefer an explicit import, but I guess that's just not how this works. A 1-line comment seems cost-effective.

File: packages/zoe/src/types.js
Comment (katelynsills) [inline packages/zoe/src/types.js]: I think Chris is thinking this won't be a promise

**Review by katelynsills — APPROVED** [review]
Some minor comments and changes, but I think most of these are comments that were also made on Chris' Zoe Split PR

File: packages/zoe/src/types.js
Comment (michaelfig) [reply packages/zoe/src/types.js]: @Chris-Hibbert I've changed this just to `Object<string,function>`, and hope your code will mesh with this definition by the time you merge to master.

File: packages/notifier/src/notifier.js
Comment (Chris-Hibbert) [inline packages/notifier/src/notifier.js]: When the state has been resolved, there should not be a new updateHandle.
Yes, that means that the declaration on line 15 should change to `| undefined`.

File: packages/zoe/src/contractFacet.js
Comment (Chris-Hibbert) [inline packages/zoe/src/contractFacet.js]: innerZoe has been renamed to zoeForZcf

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]
This PR is blocked on my changes, so I'm going to go back to finishing the zoe split rather than work on this now. I'll take another look later.

File: packages/notifier/src/notifier.js
Comment (michaelfig) [reply packages/notifier/src/notifier.js]: Got it.  Thanks!

File: packages/zoe/src/types.js
Comment (michaelfig) [reply packages/zoe/src/types.js]: Not anything to worry about: my PR won't pass tests until it meshes.

File: packages/ERTP/src/issuer.js
Comment (michaelfig) [reply packages/ERTP/src/issuer.js:20]: Highlight the type `Issuer` and hit F12 in VSCode/Codium.

Comment (dckc) [reply packages/ERTP/src/issuer.js:20]: A 1-line comment still seems cost-effective for readers not using an IDE... for example, browsing via github.

Comment (michaelfig) [reply packages/ERTP/src/issuer.js:20]: How would you word this/where you'd like to put this?  In every contract?  In all the Zoe/ERTP implementation files?

If you have a concrete suggestion, I'd be happy to implement it.

Comment (dckc) [reply packages/ERTP/src/issuer.js:20]: The `@returns {IssuerKit}` bit of jsdoc implicitly refers to `types.js`. I suggest making that explicit:

```
@returns {IssuerKit} see types.js
```

It seems particularly relevant in this case where dozens of lines of documentation were moved elsewhere without a trace. But ideally, I'd like the 1st such implicit reference in each file made explicit.

I guess another possibility is a paragraph or section in CONTRIBUTING.md about the types.js idiom.

Comment (michaelfig) [reply packages/ERTP/src/issuer.js:20]: Okay.  Thinking it over, I'll do the explicit import (`import './types'`).  That will reduce drift from what the file actually uses and any renamings of types files.

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]
This looks really good. Many more of the types are visible in WebStorm. I'll file some bugs with JetBrains and see if I can get them to improve things.

File: packages/zoe/src/types.js
Comment (Chris-Hibbert) [inline packages/zoe/src/types.js]: rather than using `omit` on the record, this should be changed to have the field name be instanceHandle.

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]

File: packages/zoe/src/contractFacet.js
Comment (michaelfig) [reply packages/zoe/src/contractFacet.js:346]: There is some pretty serious confusion about the public InstanceRecord (`handle` property) versus what I'm now calling the InternalInstanceData (`instanceHandle` property).

The table.js implementation always adds a `handle` property to records it creates, and the public Zoe/Zcf APIs expose InstanceRecords with that property.  Any `instanceHandle` is filtered out when we do so.

But, internal between Zoe and Zcf, we have `instanceHandle` instead of `handle`.  That difference seems gratuitous, and we should decide what to do about it, since at least `mintAndSellNFT.js` depends on having access to a field of the InstanceRecord called `handle`.

File: packages/zoe/src/types.js
Comment (michaelfig) [reply packages/zoe/src/types.js]: As above.

File: packages/zoe/src/contractFacet.js
Comment (michaelfig) [reply packages/zoe/src/contractFacet.js:346]: > which is where the name of the key is declared.

That key name is just for debugging the WeakMap.  It isn't actually used in the record.

Comment (michaelfig) [reply packages/zoe/src/contractFacet.js:346]: All the tests I could find destructure the results of `getInstanceRecord()` and reference the `handle` property.

Comment (Chris-Hibbert) [reply packages/zoe/src/contractFacet.js:346]: in [`zoe/test/swingsetTests/zoe/vat-bob.js`,](https://github.com/Agoric/agoric-sdk/blob/8078ef07fbe9036a84b45bdb3facb5c33cf3fc71/packages/zoe/test/swingsetTests/zoe/vat-bob.js#L511) I see
``` js
      const { installationHandle, issuerKeywordRecord, terms } = await E(
        zoe,
      ).getInstanceRecord(inviteValue[0].instanceHandle);
```

in [`zoe/test/unitTests/contracts/tests-atomicSwap.js`](https://github.com/Agoric/agoric-sdk/blob/8078ef07fbe9036a84b45bdb3facb5c33cf3fc71/packages/zoe/test/unitTests/contracts/test-atomicSwap.js#L79), I see
``` js
  const {
    installationHandle: bobInstallationId,
    issuerKeywordRecord: bobIssuers,
  } = zoe.getInstanceRecord(bobInviteValue.instanceHandle);
```

What am I missing?

Comment (Chris-Hibbert) [reply packages/zoe/src/contractFacet.js:346]: Oh, wait. I was supposed to be looking for the instanceHandle there.  Woops.  Thanks for catching me on that.

Comment (Chris-Hibbert) [reply packages/zoe/src/contractFacet.js]: The `instanceHandle` in `instanceData` should be renamed to `handle`. I'll bet there are other cascading changes. Do you want to make those changes in this PR (I could take care of that if you like) or I could do it as a separate thing.

Comment (Chris-Hibbert) [reply packages/zoe/src/contractFacet.js:346]: > But, internal between Zoe and Zcf, we have instanceHandle instead of handle. That difference seems gratuitous, and we should decide what to do about it, since at least mintAndSellNFT.js depends on having access to a field of the InstanceRecord called handle.

Yes, the difference is gratuitous and should go away.  `instanceHandle` should be renamed to `handle` as it's put in the `instanceData` record, and should retain that name in the instanceRecord passed around within contractFacet. I think the only change is that `visibleInstanceRecordFields()` should expect `handle` rather than `instanceHandle`. 

Sorry for the confusion I added. I think it's a result of Kate's asking me to document the debugging names of the key fields in state.js, which made me think the keys had those names on retrieval, which I now see was erroneous.

Once I got that in my head, I seem to have been blind to evidence to the contrary.

Comment (michaelfig) [reply packages/zoe/src/contractFacet.js:346]: Naming the `makeTable` parameter `key` was misleading.  So, I've changed it to `handleDebugName`, and defaulted it to `Handle`.

Don't worry about misunderstanding the functionality (I've made several such mistakes in the process of creating this PR): that's exactly the kind of thing that adding types helps shed light on, so consider this a useful learning experience for all of us.

**Comment by michaelfig** [comment]
I've done what I came for, and tests pass.  Ready for review.

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]
Woohoo!   Thanks, Michael, all these updates are going to make programming in this language a lot nicer.

File: packages/zoe/src/contractFacet.js
Comment (michaelfig) [reply packages/zoe/src/contractFacet.js]: Thx, I'll update the types accordingly.

File: packages/zoe/src/contractSupport/bondingCurves.js
Comment (michaelfig) [reply packages/zoe/src/contractSupport/bondingCurves.js:21]: Actually makes it into a default value rather than not parsing.

File: packages/zoe/src/rightsConservation.js
Comment (michaelfig) [reply packages/zoe/src/rightsConservation.js:37]: No.  Will update.

File: packages/zoe/src/contracts/multipoolAutoswap.js
Comment (michaelfig) [reply packages/zoe/src/contracts/multipoolAutoswap.js:124]: Agreed, good idea.

File: packages/zoe/src/contractFacet.js
Comment (michaelfig) [reply packages/zoe/src/contractFacet.js:286]: I did this entirely to appease the type checker so that I could dereference the subproperties.  I'd prefer not to go further, but yes, we could do more sanity checking.

File: packages/zoe/src/contractFacet.js
Comment (michaelfig) [reply packages/zoe/src/contractFacet.js]: That revealed a type error at:
```js
    getVatAdmin: instanceRecord.adminNode,
```
I think we wanted:
```js
    getVatAdmin: () => instanceRecord.adminNode,
```

File: packages/zoe/src/rightsConservation.js
Comment (michaelfig) [reply packages/zoe/src/rightsConservation.js:37]: Actually, it looks like `Stores` (made by `makeStore`) are passed in.  Do you see somewhere where it's an array?

Comment (Chris-Hibbert) [reply packages/zoe/src/rightsConservation.js:37]: you're right. I hadn't noticed that sumByBrand() returns Stores now.

**Review by Chris-Hibbert — APPROVED** [review]

**Pull request merged**
